lib/pbench/server/cache_manager.py

```diff
@@ -192,29 +181,155 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class LockRef:
+    """Keep track of a cache lock passed off to a caller"""
+
+    def __init__(self, lock: Path):
+        """Initialize a lock reference
+
+        The lock file is opened in "w+" mode, which is "write update": unlike
+        "r+", this creates the file if it doesn't already exist, but still
+        allows lock conversion between LOCK_EX and LOCK_SH.
+
+        Args:
+            lock: the path of a lock file
+        """
+        self.lock = lock.open("w+")
+        self.locked = False
+        self.exclusive = False
+
+    def acquire(self, exclusive: bool = False, wait: bool = True) -> "LockRef":
+        """Acquire the lock
+
+        Args:
+            exclusive: lock for exclusive access
+            wait: wait for lock
+
+        Raises:
+            OSError (EAGAIN or EACCES): wait=False and the lock is already
+                owned
+
+        Returns:
+            The lockref, so this can be chained with the constructor
+        """
+
+        self.exclusive = exclusive
+        cmd = fcntl.LOCK_EX if exclusive else fcntl.LOCK_SH
+        if not wait:
+            cmd |= fcntl.LOCK_NB
+        fcntl.lockf(self.lock, cmd)
+        self.locked = True
+        return self
+
+    def release(self):
+        """Release the lock and close the lock file"""
+
+        if not self.locked:
+            return
+        try:
+            fcntl.lockf(self.lock, fcntl.LOCK_UN)
+            self.lock.close()
+        finally:
+            # Release our reference to the lock file so that the
+            # object can be reclaimed.
+            self.locked = False
+            self.exclusive = False
+            self.lock = None
+
+    def upgrade(self):
+        """Upgrade a shared lock to exclusive"""
+        if not self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_EX)
+            self.exclusive = True
+
+    def downgrade(self):
+        """Downgrade an exclusive lock to shared"""
+        if self.exclusive:
+            fcntl.lockf(self.lock, fcntl.LOCK_SH)
+            self.exclusive = False
+
+
+class LockManager:
+    def __init__(self, lock: Path, exclusive: bool = False, wait: bool = True):
+        self.lock = LockRef(lock)
+        self.exclusive = exclusive
+        self.wait = wait
+        self.unlock = True
+
+    def keep(self) -> "LockManager":
+        """Tell the context manager not to unlock on exit"""
+        self.unlock = False
+        return self
+
+    def release(self):
+        """Release manually if necessary"""
+        self.lock.release()
```

Comment:
Should this set `self.unlock` to `False`?

Reply:
It doesn't matter, since `LockRef.release()` is protected. Normally, `LockManager.release()` only makes sense outside the context manager anyway, when we exited with `keep`, so nobody's going to be looking. (I use this in `stream` to defer to unlock/close until post-response time via `Inventory.close`.)